src/workerd/api/sockets.h

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include <workerd/jsg/jsg.h>
+#include "http.h"
+
+
+namespace workerd::api {
+
+class PipelinedAsyncIoStream final : public kj::AsyncIoStream, public kj::Refcounted {
```

Comment:
Is this significantly different from `PromisedAsyncIoStream` in KJ? If so, please explain how in the comments. If not, let's reuse the existing implementation. I'm specifically referring to this function in `kj/async-io.h`:

```
Own<AsyncIoStream> newPromisedStream(Promise<Own<AsyncIoStream>> promise);
```

Reply:
Wasn't aware of this (and I did actually ask at least one person whether it exists and they said to implement it). But yeah, I'll refactor to use this in https://github.com/cloudflare/workerd/pull/205.